texar/modules/encoders/xlnet_encoders.py

```diff
@@ -0,0 +1,606 @@
+            "vocab_size": 32000,
+            "max_seq_len": 512,
+            '@no_typecheck': ['pretrained_model_name']
+        }
+
+    def param_groups(self, lr=None, lr_layer_scale=1.0,
+                     decay_base_params=False):
+        r"""Create parameter groups for optimizers. When
+        :attr:`lr_layer_decay_rate` is not 1.0, parameters from each layer form
+        separate groups with different base learning rates.
+
+        This method should be called before applying gradients to the variables
+        through the optimizer. Particularly, after calling the optimizer's
+        `compute_gradients` method, the user can call this method to get
+        variable-specific learning rates for the network. The gradients for each
+        variables can then be scaled accordingly. These scaled gradients are
+        finally applied by calling optimizer's `apply_gradients` method.
+
+        Args:
+            lr (float): The learning rate. Can be omitted if
+                :attr:`lr_layer_decay_rate` is 1.0.
+            lr_layer_scale (float): Per-layer LR scaling rate. The `i`-th layer
+                will be scaled by `lr_layer_scale ^ (num_layers - i - 1)`.
+            decay_base_params (bool): If `True`, treat non-layer parameters
+                (e.g. embeddings) as if they're in layer 0. If `False`, these
+                parameters are not scaled.
+
+        Returns: A dict mapping tensorflow variables to their learning rates.
+        """
+        vars_to_learning_rates = {}
+        if lr_layer_scale != 1.0:
+            if lr is None:
+                raise ValueError(
+                    "lr must be specified when lr_layer_decay_rate is not 1.0")
+
+            num_layers = self._hparams.num_layers
+            scope = self.variable_scope.name
+            base_var_names = ['r_w_bias', 'r_r_bias', 'word_embedder']
+
+            if self._hparams.use_segments:
+                base_var_names.extend(['r_s_bias', 'seg_embed'])
+
+            for var in base_var_names:
+                tf_variable = tf.trainable_variables(scope=scope + "/" + var)[0]
+                vars_to_learning_rates[tf_variable] = \
+                    lr * (lr_layer_scale ** num_layers if decay_base_params
+                          else 1.0)
+
+            for idx in range(num_layers):
+                decay_rate = lr_layer_scale ** (num_layers - idx - 1)
+                layer_variables = tf.trainable_variables(
+                    scope=scope + "/" + "layer_{}".format(idx))
+                for variable in layer_variables:
+                    vars_to_learning_rates[variable] = lr * decay_rate
+        else:
+            for variable in self.trainable_variables:
+                vars_to_learning_rates[variable] = lr
+
+        return vars_to_learning_rates
+
+    @property
+    def output_size(self):
+        r"""Return the size of the undivided hidden dimension of the network.
+        """
+        return self._hparams.hidden_dim
```

Comment:
`output_size` is defined to be "The final dimension(s) of :meth:`forward` output tensor(s)" https://github.com/asyml/texar-pytorch/blob/master/texar/torch/module_base.py#L95

Now the `forward` returns two tensors, why is the output_size here a single size? (Same question applies to the PyTorch version)

Reply:
I discussed with Zecong about this. At other places, `output_size` returns an int denoting the size of the network. If we instead return sizes of both tensors of `_build`, the user would have to handle this case differently which may be a bit of a trouble for the user. It is important to note that user actually cares about the size of `output` tensor returned by `_build`. I have added a note to indicate that this method returns the size of the `output` tensor.

Reply:
Then the docstring "The final dimension(s) of :meth:forward output tensor(s)" should be updated. It's not exactly the size(s) of forward outputs, but instead more of case-by-case? E.g., sometimes it's the dimensions of some of the output tensors. 

And, what does "final dimension(s)" mean? 

Since it's case-by-case, all `output_size` should explicitly and clearly explain what it is in the particular classes.

Reply:
I will create an issue in `texar-pytorch` to track and fix this. In `texar-tf`, I just checked, other modules don't actually have `output_size` method. I happened to have copied it from `texar-pytorch`. So we can either keep it or remove it altogether in `texar-tf`.

Reply:
Let's keep it here. To make the docstring absolutely clear, we can probably say: 

```
The last dimension of the encoder output.
	
Note: The :meth:`_build` returns two tensors of shapes 
`[batch_size, max_time, hidden_dim]` and `[batch_size, cache_len, hidden_dim]`.
 output_size here equals `hidden_dim`
```

Reply:
For class properties (with `@property` decorator), the docstring usually does not say `Return xxx` cuz it's not  used as a function

Reply:
Fixed in 1ffca93